crates/spirv-std/src/image.rs

```diff
@@ -0,0 +1,713 @@
+//! Image types
+
+// Rustfmt formats long marker trait impls over multiple lines which makes them
+// harder to read.
+#[rustfmt::skip]
+mod params;
+
+pub use self::params::{ImageCoordinate, SampleType};
+pub use crate::macros::Image;
+pub use spirv_types::image_params::{
+    AccessQualifier, Arrayed, Dimensionality, ImageDepth, ImageFormat, Multisampled, Sampled,
+};
+
+use crate::{float::Float, integer::Integer, vector::Vector, Sampler};
+
+/// Re-export of primitive types to ensure the `Image` proc macro always points
+/// to the right type.
+pub mod __private {
+    pub use {f32, f64, i16, i32, i64, i8, u16, u32, u64, u8};
+}
+
+/// An opaque image type. Corresponds to `OpTypeImage`.
+#[spirv(generic_image_type)]
+#[derive(Copy, Clone)]
+pub struct Image<
+    SampledType: SampleType<FORMAT>,
+    const DIM: Dimensionality,
+    const DEPTH: ImageDepth,
+    const ARRAYED: Arrayed,
+    const MULTISAMPLED: Multisampled,
+    const SAMPLED: Sampled,
+    const FORMAT: ImageFormat,
+    const ACCESS_QUALIFIER: Option<AccessQualifier>,
+> {
+    _x: u32,
+    _marker: core::marker::PhantomData<SampledType>,
+}
+
+impl<
+        SampledType: SampleType<FORMAT>,
+        const DIM: Dimensionality,
+        const DEPTH: ImageDepth,
+        const ARRAYED: Arrayed,
+        const MULTISAMPLED: Multisampled,
+        const SAMPLED: Sampled,
+        const FORMAT: ImageFormat,
+        const ACCESS_QUALIFIER: Option<AccessQualifier>,
+    > Image<SampledType, DIM, DEPTH, ARRAYED, MULTISAMPLED, SAMPLED, FORMAT, ACCESS_QUALIFIER>
+{
+    /// Sample texels at `coord` from the image using `sampler`.
+    #[crate::macros::gpu_only]
+    pub fn sample<F, V>(&self, sampler: Sampler, coord: impl ImageCoordinate<F, DIM, ARRAYED>) -> V
+    where
+        F: Float,
+        V: Vector<SampledType, 4>,
+    {
+        unsafe {
+            let mut result = Default::default();
+            asm!(
+                "%typeSampledImage = OpTypeSampledImage typeof*{1}",
+                "%image = OpLoad typeof*{1} {1}",
+                "%sampler = OpLoad typeof*{2} {2}",
+                "%coord = OpLoad typeof*{3} {3}",
+                "%sampledImage = OpSampledImage %typeSampledImage %image %sampler",
+                "%result = OpImageSampleImplicitLod typeof*{0} %sampledImage %coord",
```

Comment:
From the spec:

> OpImageSampleImplicitLod
> The MS operand of the underlying OpTypeImage must be 0.

Might be nice to encode that, as well as all the other restrictions here, but that can be done in a follow-up.